cmd/utils/cmd.go

```diff
@@ -224,6 +229,88 @@ func ImportChain(chain *core.BlockChain, fn string) error {
 	return nil
 }
 
+func readList(filename string) ([]string, error) {
+	b, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	return strings.Split(string(b), "\n"), nil
+}
+
+// ImportHistory imports Era1 files containing historical block information,
+// starting from genesis.
+func ImportHistory(chain *core.BlockChain, db ethdb.Database, dir string, network string) error {
+	if chain.CurrentSnapBlock().Number.BitLen() != 0 {
+		return fmt.Errorf("history import only supported when starting from genesis")
+	}
+	entries, err := era.ReadDir(dir, network)
+	if err != nil {
+		return fmt.Errorf("error reading %s: %w", dir, err)
+	}
+	checksums, err := readList(path.Join(dir, "checksums.txt"))
+	if err != nil {
+		return fmt.Errorf("unable to read checksums.txt: %w", err)
+	}
+	if len(checksums) <= len(entries) {
+		return fmt.Errorf("invalid checksums.txt: attempting to import more era1 files than have checksums")
+	}
+	var (
+		start    = time.Now()
+		reported = time.Now()
+		imported = 0
+		forker   = core.NewForkChoice(chain, nil)
+	)
+	for i, filename := range entries {
+		// Read entire Era1 to memory. Max historical Era1 is around
+		// 600MB. This is a lot to load at once, but it speeds up the
+		// import substantially.
+		b, err := os.ReadFile(path.Join(dir, filename))
+		if err != nil {
+			return fmt.Errorf("unable to open era: %w", err)
+		}
+
+		if have, want := common.Hash(sha256.Sum256(b)).Hex(), checksums[i]; have != want {
+			return fmt.Errorf("checksum mismatch: have %s, want %s", have, want)
+		}
+
+		// Import all block data from Era1.
+		r := era.NewReader(bytes.NewReader(b))
+		for j := 0; ; j += 1 {
+			n := i*era.MaxEra1Size + j
+			b, r, err := r.Read()
+			if err == io.EOF {
+				break
+			} else if err != nil {
+				return fmt.Errorf("error reading block %d: %w", n, err)
+			} else if b.Number().BitLen() == 0 {
+				continue // skip genesis
+			}
+			if status, err := chain.HeaderChain().InsertHeaderChain([]*types.Header{b.Header()}, start, forker); err != nil {
+				return fmt.Errorf("error inserting header %d: %w", n, err)
+			} else if status != core.CanonStatTy {
+				return fmt.Errorf("error inserting header %d, not canon: %v", n, status)
+			}
+			if _, err := chain.InsertReceiptChain([]*types.Block{b}, []types.Receipts{r}, 2^64-1); err != nil {
+				return fmt.Errorf("error inserting body %d: %w", n, err)
+			}
+			imported += 1
+
+			// Give the user some feedback that something is happening.
+			if time.Since(reported) >= 8*time.Second {
+				log.Info("Importing Era files", "head", n, "imported", imported, "elapsed", common.PrettyDuration(time.Since(start)))
+				imported = 0
+				reported = time.Now()
+			}
+		}
+	}
+
+	// Manually update the head snap block so that when user restarts geth,
+	// it is able to begin syncing from the last imported block.
+	rawdb.WriteHeadFastBlockHash(db, chain.CurrentHeader().Hash())
```

Comment:
Shouldn't this be written in blockchain/InsertReceiptChain already? (I might be wrong)

Reply:
Yeah you're right, it is now, at some point I was doing something a bit different and had to write in manually.